include/ocpp/v201/ocpp_types.hpp

```diff
@@ -959,6 +969,86 @@ void from_json(const json& j, Firmware& k);
+/// \param profile the charging profile
+/// \return a list of the start times of the profile
+std::vector<DateTime> calculate_start(const DateTime& now, const DateTime& end,
+                                      const std::optional<DateTime>& session_start, const ChargingProfile& profile);
+
+/// \brief calculate the start times for the schedule period
+/// \param now the current date and time
+/// \param end the end of the composite schedule
+/// \param session_start optional when the charging session started
+/// \param profile the charging profile
+/// \param period_index the schedule period index
+/// \return the list of start times
+std::vector<period_entry_t> calculate_profile_entry(const DateTime& now, const DateTime& end,
+                                                    const std::optional<DateTime>& session_start,
+                                                    const ChargingProfile& profile, std::uint8_t period_index);
+
+/// \brief generate an ordered list of valid schedule periods for the profile
+/// \param now the current date and time
+/// \param end ignore entries beyond this date and time (i.e. that start after end)
+/// \param session_start optional when the charging session started
+/// \param profile the charging profile
+/// \return a list of profile periods with calculated date & time start and end times
+/// \note it is valid for there to be gaps (for recurring profiles)
+std::vector<period_entry_t> calculate_profile(const DateTime& now, const DateTime& end,
+                                              const std::optional<DateTime>& session_start,
+                                              const ChargingProfile& profile);
+
+/// \brief calculate the composite schedule for the list of periods
+/// \param combined_schedules the list of periods to build into the schedule
+/// \param now the start of the composite schedule
+/// \param end the end (i.e. duration) of the composite schedule
+/// \param charging_rate_unit the units to use (defaults to Amps)
+/// \return the calculated composite schedule
+CompositeSchedule calculate_composite_schedule(std::vector<period_entry_t>& combined_schedules, const DateTime& now,
+                                               const DateTime& end,
+                                               std::optional<ChargingRateUnitEnum> charging_rate_unit);
+
+/// \brief calculate the composite combined schedule
+/// \param charge_point_max the composite schedule for ChargePointMax profiles
+/// \param tx_default the composite schedule for TxDefault profiles
+/// \param tx the composite schedule for Tx profiles
+/// \return the calculated combined composite schedule
+/// \note all composite schedules must have the same units configured
+CompositeSchedule calculate_composite_schedule(const CompositeSchedule& charging_station_max,
+                                               const CompositeSchedule& tx_default, const CompositeSchedule& tx);
```

Comment:
I think we're overloading on this name a bit too much; I wasn't sure what we were calling when we called `calculate_composite_schedule()` within `SmartChargingHandler::calculate_composite_schedule()`, and I felt especially lost realizing that the function took two completely different sets of arguments.

Maybe we could have:

* `calculate_composite_schedule_for_periods()`
* `calculate_combined_composite_schedule()`

Also, is there any particular reason these are separate from `SmartChargingHandler`?

Reply:
I agree. The reason I decided not to change it is that it matches the usage in the 1.6 version of the code. I'm trying to keep them in alignment as much as possible.